packages/experiment-core/src/api/stream-api.ts

```diff
@@ -0,0 +1,194 @@
+const MAX_CONN_MS_MAX = 18 * 60 * 1000; // Max of the above timeout.
+const KEEP_ALIVE_INTERVAL = (15 + 2) * 1000; // 15 seconds plus 2 seconds grace period. // 0 or neg value disables keep alive.
+const KEEP_ALIVE_DATA = ' ';
+
+export type StreamOptions = {
+  libraryName: string;
+  libraryVersion: string;
+};
+export type StreamOnUpdateCallback = (data: string) => unknown;
+export type StreamOnErrorCallback = (err: StreamErrorEvent) => unknown;
+
+export interface StreamApi {
+  /**
+   * Initiate a connection. If an existing connection exists, it does nothing.
+   */
+  connect(options?: StreamOptions): void;
+  /**
+   * Close a connection. If there is no existing connection, it does nothing.
+   */
+  close(): void;
+  /**
+   * Any message will be sent to this callback.
+   */
+  onUpdate?: StreamOnUpdateCallback;
+  /**
+   * Any error, including connection errors, will be sent to this callback.
+   */
+  onError?: StreamOnErrorCallback;
+}
+
+/**
+ * This class handles connecting to an server-side event source.
+ * It handles keep alives from server, automatically disconnect and reconnect after a set random interval.
+ * It will propagate any error to onError. It will not handle any error or retries.
+ * (automatic disconnect does not count as error, but if any reconnect errors will propagate).
+ */
+export class SdkStreamApi implements StreamApi {
+  private eventSource: StreamEventSource | undefined;
+  private reconnectionTimeout?: NodeJS.Timeout;
+  private initConnTimeout?: NodeJS.Timeout;
+  private keepAliveTimeout?: NodeJS.Timeout;
+
+  public onUpdate?: StreamOnUpdateCallback;
+  public onError?: StreamOnErrorCallback;
+
+  private readonly deploymentKey: string;
+  private readonly serverUrl: string;
+  private readonly eventSourceClass: StreamEventSourceClass;
+  private streamConnTimeoutMillis: number; // Timeout for connecting the stream. Aka, http connection timeout.
+
+  constructor(
+    deploymentKey: string,
+    serverUrl: string,
+    eventSourceClass: StreamEventSourceClass,
+    streamConnTimeoutMillis: number = DEFAULT_INITIAL_CONN_TIMEOUT, // Timeout for connecting the stream. Aka, http connection timeout.
+  ) {
+    this.deploymentKey = deploymentKey;
+    this.serverUrl = serverUrl;
+    this.eventSourceClass = eventSourceClass;
+    this.streamConnTimeoutMillis = streamConnTimeoutMillis;
+  }
+
+  public async connect(options?: StreamOptions) {
+    if (this.eventSource) {
+      return;
+    }
+
+    const headers: Record<string, string> = {
+      Authorization: `Api-Key ${this.deploymentKey}`,
+    };
+    if (options?.libraryName && options?.libraryVersion) {
+      headers[
+        'X-Amp-Exp-Library'
+      ] = `${options.libraryName}/${options.libraryVersion}`;
+    }
+
+    // Create connection. It starts connection on new.
+    const es = new this.eventSourceClass(this.serverUrl, {
```

Comment:
I dont think `EventSourceClass` really describes what this is.

Maybe something like `StreamEventSourceConnectionFactory` and `StreamEventSourceConnection` rather than `StreamEventSource` and `StreamEventSourceClass`

That is just a thought. My suggestions are a little wordy but my point is that really the "class" is creating the connection which is used until it is closed. "Factory" might not be the right suffix here, and I'm still not sold on shoehorning in `new` instead of using an explicit interface with a function that creates the connection using the underlying implementation.

Reply:
I was torned between factory pattern or just passing a class. Chose passing a class to keep the interface similar to other APIs, but factory pattern does makes more sense here.